### PR TITLE
Drop support for hapi < v20

### PR DESCRIPTION
**Pull request by mrickard**
## Proposed Release Notes

- Dropped instrumentation for hapi < v20.0.0. Those versions have been deprecated by Hapi.
- Dropped tests for hapi < v20.0.0.

## Links

* https://issues.newrelic.com/browse/NR-33405

## Details
All versions < 20.0.0 have been deprecated by Hapi, due to security issues. See: https://www.npmjs.com/package/@hapi/hapi versions < 20 and all versions of https://www.npmjs.com/package/hapi.

### CODE REVIEW

**Comment by bizob2828** [comment]
@mrickard can you please keep the PR template and fill it out accordingly?  We derive release notes from that and also link github/jira issues that way

**Comment by mrickard** [comment]
@bizob2828 OK--I'll update with the template.

File: test/versioned/hapi/hapi-v20/package.json [hunk 1]
```diff
@@ -1,16 +1,16 @@
...
       "engines": {
-        "node": ">=12 < 16"
+        "node": ">=14"
       },
       "dependencies": {
         "ejs": "2.5.5",
-        "hapi": "^17.0.0",
-        "vision": "5.1"
+        "@hapi/hapi": ">=20.1.2",
+        "vision": "latest"
```
Comment (michaelgoin) [inline test/versioned/hapi/hapi-v20/package.json:13]: Noting this has changed from a single version to "latest." That is fine, it didn't used to be possible but is now.

I did take a quick check and `vision`, without a namespace, is no-longer maintained. I'm not sure we even have support for `@hapi/vision`. I'm also not sure if that instrumentation is needed anymore for Hapi. We may want to capture a story to dig into that.

File: test/versioned/hapi/hapi-v20/package.json [hunk 2]
```diff
@@ -1,16 +1,16 @@
 {
-  "name": "hapi-v17-tests",
+  "name": "hapi-v20-tests",
```
Comment (michaelgoin) [inline test/versioned/hapi/hapi-v20/package.json:2]: Looks like we've dropped hapi-post18 with hapi-v20 as the replacement for all tests. It is interesting that we've ported over some of the tests from 17 which were defined explicitly in post-18, which was the previous coverage for 20. Unsure if there were any differences there to be aware of or not in those test files.

**Review by michaelgoin — COMMENTED** [review]
With this change, we fully don't support the `hapi` module that is not namespaced. 

As such, we should remove from `'hapi': { type: MODULE_TYPE.WEB_FRAMEWORK },` from instrumentations and can cleanup anything related to that. We should at least remove the `hapi` reference in instrumentations prior to releasing. That can be done as a part of this PR or separately (whether right away with current story or added as a new story).

If you'd like to do separately, I can approve this PR. Leaving as a comment for now as not to have that get lost in the mix.

I think we should do a second and maybe third pass to clean up the Hapi stuff. Some of that can happen post-release, as desired. But seems like we can simplify down the instrumentation and tests potentially further. Although, I do see a v21 is in beta so we may want to hold off on some of that just to make sure we won't need the forking. That being said, I don't want to hold up the first pass on the PR to get it perfect. It will be easier to see the various impacts with this landed.

File: lib/instrumentation/hapi.js [hunk 3]
```diff
@@ -27,16 +27,7 @@ module.exports = function initialize(agent, hapi, moduleName, shim) {
...
-    }
-
-    // See if we can find the plugin class. This should be the super class of
-    // Server and will cover more scenarios.
-    const Plugin = hapi.Server.super_
-    if (_isPluginClass(Plugin)) {
-      wrapServer(shim, Plugin)
-    } else {
-      wrapServer(shim, hapi.Server)
+      return require('./hapi/hapi-20')(agent, hapi, moduleName, shim)
```
[the remark this answers is not in the document]
Comment (mrickard) [reply lib/instrumentation/hapi.js:30]: I'd be in favor of removing the version from these file names, and then if version-forking is required in the future, we can revisit then. 

The versioned-test directory structure isn't changing, apart from getting that version number out of the file names. (It turns out that Vision's making assumptions about directory structure.)

File: test/versioned/hapi/hapi-v20/package.json (hunk 1 above)
Comment (mrickard) [reply test/versioned/hapi/hapi-v20/package.json:13]: @michaelgoin I'll add a story for that.

File: test/versioned/hapi/hapi-v20/capture-params.tap.js [hunk 4]
```diff
@@ -9,7 +9,7 @@ const DESTINATIONS = require('../../../../lib/config/attribute-filter').DESTINAT
 const tap = require('tap')
 const request = require('request')
 const helper = require('../../../lib/agent_helper')
-const utils = require('./hapi-17-utils')
+const utils = require('./hapi-20-utils')
```
[the remark this answers is not in the document]
Comment (mrickard) [reply test/versioned/hapi/hapi-v20/capture-params.tap.js]: Updated to remove the version number.

File: test/versioned/hapi/hapi-v20/package.json (hunk 2 above)
Comment (mrickard) [reply test/versioned/hapi/hapi-v20/package.json]: > With this change, we fully don't support the `hapi` module that is not namespaced.
> 
> As such, we should remove from `'hapi': { type: MODULE_TYPE.WEB_FRAMEWORK },` from instrumentations and can cleanup anything related to that. 

I've removed that from instrumentations. The `hapi` and `@hapi/hapi` instrumentation structure is pretty DRY: the namespaced one points back at the other.  I could refactor that in this PR to consolidate it all under the namespaced directory.

**Review by michaelgoin — APPROVED** [review]
👍

**Pull request merged**